src/miner_val_heartbeat.erl

```diff
@@ -0,0 +1,129 @@
+                        chain = Chain}}
+    end.
+
+handle_call(_Request, _From, State) ->
+    lager:warning("unexpected call ~p from ~p", [_Request, _From]),
+    Reply = ok,
+    {reply, Reply, State}.
+
+handle_cast(_Msg, State) ->
+    lager:warning("unexpected cast ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info({blockchain_event, {add_block, _Hash, _Sync, _Ledger}},
+            #state{txn_status = waiting, txn_wait = Wait} = State) ->
+    case Wait of
+        1 ->
+            {noreply, State#state{txn_status = ready, txn_wait = 10}};
+        N ->
+            {noreply, State#state{txn_wait = N - 1}}
+    end;
+handle_info({blockchain_event, {add_block, Hash, Sync, _Ledger}},
+            #state{address = Address, sigfun = SigFun} = State) ->
+    Ledger = blockchain:ledger(State#state.chain),
+    case blockchain:config(?election_version, Ledger) of
+        {ok, V} when V >= 5 ->
+            {ok, HBInterval} = blockchain:config(?validator_liveness_interval, Ledger),
+            {ok, Block} = blockchain:get_block(Hash, State#state.chain),
+            Height = blockchain_block:height(Block),
+            %% heartbeat server needs to be able to run on an unstaked validator
+            case blockchain_ledger_v1:get_validator(Address, Ledger) of
+                {ok, Val} ->
+                    lager:debug("getting validator for address ~p got ~p", [Address, Val]),
+                    case blockchain_ledger_validator_v1:last_heartbeat(Val) of
+                        N when (N + HBInterval) =< Height andalso (not Sync) ->
+                            %% we need to construct and submit a heartbeat txn
+                            {ok, CBMod} = blockchain_ledger_v1:config(?predicate_callback_mod, Ledger),
+                            {ok, Callback} = blockchain_ledger_v1:config(?predicate_callback_fun, Ledger),
+                            UnsignedTxn =
+                                blockchain_txn_validator_heartbeat_v1:new(Address, Height, CBMod:Callback()),
+                            Txn = blockchain_txn_validator_heartbeat_v1:sign(UnsignedTxn, SigFun),
+                            lager:debug("submitting txn ~p for val ~p ~p ~p", [Txn, Val, N, HBInterval]),
+                            Self = self(),
+                            blockchain_worker:submit_txn(Txn, fun(Res) -> Self ! {sub, Res} end),
+                            {noreply, State#state{txn_status = waiting}};
+                        _ -> {noreply, State}
+                    end;
+                {error, not_found} ->
+                    lager:debug("getting validator for address ~p not found", [Address]),
+                    {noreply, State}
+            end;
+        _ ->
+            {noreply, State}
+    end;
+%% logically it doesn't matter what the result is, we either need to start actively waiting or
+%% trying again
+handle_info({sub, _Res}, State) ->
+    {noreply, State#state{txn_status = ready, txn_wait = 10}};
```

Comment:
Would it not be better to look at the result here and also use the `validator_liveness_interval` to inform the txn_wait?

Reply:
I don't actually need to look at the response here, since if we succeeded, we won't submit another one, and if we failed, we'll just submit another one in the next block we get